core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceBranch.java

```diff
@@ -267,6 +267,10 @@ public void flush() throws SailException {
 					prepared = null;
 				}
 			}
+		} catch (SailException e) {
+			// clear changes if flush fails
+			changes.clear();
```

Comment:
Can we add a regression test of some sort that demonstrates the situation this fixes? I realize there isn't a lot of unit test coverage for this code, but perhaps it would fit in as part of the OptmisticIsolationTest suite? 

Also: should you do the same kind of thing for the `prepared` iterms (presunably inside the `prepare()` method)?

Reply:
Yes, for sure. I also think this could somehow be integrated into the OptimisticIsolationTest suite. I will take a look into this.